### PR TITLE
Fix #8797: Add OAuth2AuthenticationException to allowlist

### LINKED ISSUE #8797: OAuth2AuthenticationException should be in allowlist
**Describe the bug**
Using `spring-boot-starter-oauth2-client` with `spring-session-data-redis` and JSON serialization, when attempting an invalid authentication (for example by using the browsers back button) an `OAuth2AuthenticationException` ends up being serialized into the Redis store, which then cannot be deserialized due to the security config.

**To Reproduce**
I noticed this problem while developing an open source project of mine. The steps to reproduce this are fairly complicated and require amongst other things a Discord account and registering a [Discord bot application](https://discord.com/developers/applications). Unfortunately I have no idea how to reproduce it more minimally, without talking to a third-party OAuth2 provider. At least the steps below allow to reproduce it using a port forward service, no full deployment necessary.

Maybe the stack traces and serialized data which are posted further down this post are enough to identify the problem without having to go through these reproduction steps.

Here we go:

The current state of the project as of writing this is this commit: https://github.com/wolfiabot/Wolfia/tree/de5fb5094d8c586d2d793c78adf0cbce08d52d60
A more-or-less detailed guide to start the backend and the frontend it locally is here: https://github.com/wolfiabot/Wolfia/tree/de5fb5094d8c586d2d793c78adf0cbce08d52d60/dashboard

Once everything is running, open it in your browser.
1. Click on the login button in the top right
2. Login to Discord if necessary, and click Authorize
3. You will be redirected back to the application, which now detects you as logged in. All good so far!
4. Click the back button in the browser to go back to Discord's OAuth2 screen
5. Click Authorize again

The application throws an HTTP 500 for any requests to the backend now. Only clearing the cookies will help, as a session with the undeserializable data has been written into redis.

```
org.springframework.data.redis.serializer.SerializationException: Could not read JSON: The class with org.springframework.security.oauth2.core.OAuth2AuthenticationException and name of org.springframework.security.oauth2.core.OAuth2AuthenticationException is not whitelisted. If you believe this class is safe to deserialize, please provide an explicit mapping using Jackson annotations or by providing a Mixin. If the serialization is only done by a trusted source, you can also enable default typing. See https://github.com/spring-projects/spring-security/issues/4370 for details; nested exception is java.lang.IllegalArgumentException: The class with org.springframework.security.oauth2.core.OAuth2AuthenticationException and name of org.springframework.security.oauth2.core.OAuth2AuthenticationException is not whitelisted. If you believe this class is safe to deserialize, please provide an explicit mapping using Jackson annotations or by providing a Mixin. If the serialization is only done by a trusted source, you can also enable default typing. See https://github.com/spring-projects/spring-security/issues/4370 for details
```
The full stack trace of the exception is here:
https://wastebin.party/ojidoxomol.pl

I used [Redsmin](https://www.redsmin.com/) to look into the exact content written into redis:
image
It appears to me the offending data is in the `sessionAttr:SPRING_SECURITY_LAST_EXCEPTION` field:
https://wastebin.party/eseyinomeg.json

**Expected behavior**
Data that the framework serializes should also be deserializable by itself.
I can imagine two ways to achieve that: Either don't serialize any `OAuth2AuthenticationException`s in the first place, or whitelist them.
Also not ruling out that something with my own setup is wrong here in which case some pointers are very welcome.

**Sample**
See the reproduction steps above.

Please let me know if anything is unclear and I'll do my best to provide further information.

I have suggested two possible approaches to a fix under "Expected behaviour". I'd be happy to work on a pull request to fix this problem if someone with better understanding of the internals of the framework can give me some directions which of those (or something else entirely) would be the correct approach.

### PR DESCRIPTION
**Pull request by napstr**
Add mixins for
- OAuth2AuthenticationException
- OAuth2Error

See #8797 for more details.

### CODE REVIEW

**Comment by napstr** [comment]
@jgrandja here's my take on #8797, what do you think?


Side note: I could not run the whole test suite locally to verify nothing else breaks. Master branch is failing locally with the same error as the CI build:
```
org.springframework.security.concurrent.CurrentDelegatingSecurityContextScheduledExecutorServiceTests > invokeAnyTimeout FAILED
    java.lang.NoSuchMethodError at PowerMockitoInjectingAnnotationEngine.java:40
```

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixin.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * This mixin class is used to serialize/deserialize
+ * {@link OAuth2AuthenticationException}.
+ *
+ * @author Dennis Neufeld
+ * @since 5.4
+ * @see OAuth2AuthenticationException
+ * @see OAuth2ClientJackson2Module
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
+@JsonIgnoreProperties(ignoreUnknown = true, value = {"cause", "stackTrace"})
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixin.java:35]: Please add

```java
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
		isGetterVisibility = JsonAutoDetect.Visibility.NONE)
```

so it's consistent with the other mixins

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ErrorMixin.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * {@link org.springframework.security.oauth2.core.OAuth2AuthenticationException}.
+ *
+ * @author Dennis Neufeld
+ * @since 5.4
+ * @see OAuth2Error
+ * @see OAuth2AuthenticationExceptionMixin
+ * @see OAuth2ClientJackson2Module
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ErrorMixin.java:35]: Please add

```java
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
		isGetterVisibility = JsonAutoDetect.Visibility.NONE)
```

so it's consistent with the other mixins

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ClientJackson2Module.java [hunk 3]
```diff
@@ -101,5 +103,7 @@ public void setupModule(SetupContext context) {
 		context.setMixInAnnotations(OidcUserAuthority.class, OidcUserAuthorityMixin.class);
 		context.setMixInAnnotations(DefaultOidcUser.class, DefaultOidcUserMixin.class);
 		context.setMixInAnnotations(OAuth2AuthenticationToken.class, OAuth2AuthenticationTokenMixin.class);
+		context.setMixInAnnotations(OAuth2AuthenticationException.class, OAuth2AuthenticationExceptionMixin.class);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ClientJackson2Module.java:106]: Add `OAuth2AuthenticationExceptionMixin` and `OAuth2ErrorMixin` to class-level javadoc

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+			+ "\n    \"uri\":null"
+			+ "\n  },"
+			+ "\n  \"message\":\"[authorization_request_not_found] \","
+			+ "\n  \"suppressed\":[\"[Ljava.lang.Throwable;\",[]],"
+			+ "\n  \"localizedMessage\":\"[authorization_request_not_found] \""
+			+ "\n}";
+	// @formatter:on
+
+	@Test
+	public void serializeOAuth2AuthenticationExceptionMixinTest() throws JsonProcessingException, JSONException {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:66]: Please rename test methods to following the naming convention **methodName**When**Condition**Then**Result** - see `OAuth2AuthorizedClientMixinTests` for examples

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+			+ "\n  \"@class\": \"org.springframework.security.oauth2.core.OAuth2AuthenticationException\","
+			+ "\n  \"error\":"
+			+ "\n  {"
+			+ "\n    \"@class\":\"org.springframework.security.oauth2.core.OAuth2Error\","
+			+ "\n    \"errorCode\":\"authorization_request_not_found\","
+			+ "\n    \"description\":null,"
+			+ "\n    \"uri\":null"
+			+ "\n  },"
+			+ "\n  \"message\":\"[authorization_request_not_found] \","
+			+ "\n  \"suppressed\":[\"[Ljava.lang.Throwable;\",[]],"
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:60]: Add `suppressed` to `JsonIgnoreProperties`

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+			+ "\n  \"localizedMessage\":\"[authorization_request_not_found] \""
+			+ "\n}";
+	// @formatter:on
+
+	@Test
+	public void serializeOAuth2AuthenticationExceptionMixinTest() throws JsonProcessingException, JSONException {
+		OAuth2Error oauth2Error = new OAuth2Error("authorization_request_not_found");
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:67]: Please add a `description` and `uri` to test this fully

**Comment by napstr** [comment]
@jgrandja thanks for the detailed feedback. I think I have now addressed all your points, and expanded the test cases following the example set by other mixin tests.
Oh, also, using a local snapshot of this branch, I can confirm this solves the issue in my application.

**Comment by jgrandja** [comment]
Thanks for the updates @napstr ! This is now in master and backported to 5.3.x

**Pull request closed without merging**